marlowe-playground-client/src/Marlowe/Linter.purs

```diff
@@ -138,198 +147,427 @@ termToRange a { row, column } =
   in
-    lintValue newState value
+    do
+      _ <- CMS.modify (over _holes gatherHoles)
+      pure (ValueSimp pos false t)
+
+lintValue env (Term (Constant (Term v pos2)) pos) = do
+  pure (ConstantSimp pos false v)
+
+lintValue env t@(Term (Constant h@(Hole _ _ _)) pos) = do
+  _ <- CMS.modify (over _holes (insertHole h))
+  pure (ValueSimp pos false t)
+
+lintValue env t@(Term (NegValue a) pos) = do
+  sa <- lintValue env a
+  case sa of
+    ConstantSimp _ _ c1 -> pure (ConstantSimp pos true (-c1))
+    _ -> do
+      markSimplification constToVal SimplifiableValue a sa
+      pure (ValueSimp pos false t)
+
+lintValue env t@(Term (AddValue a b) pos) = do
+  sa <- lintValue env a
+  sb <- lintValue env b
+  case sa /\ sb of
+    (ConstantSimp _ _ v1 /\ ConstantSimp _ _ v2) -> pure (ConstantSimp pos true (v1 + v2))
+    (ConstantSimp _ _ v /\ _) ->
+      if (v == zero) then
+        pure (simplifyTo sb pos)
+      else
+        defaultCase sa sb
+    (_ /\ ConstantSimp _ _ v) ->
+      if (v == zero) then
+        pure (simplifyTo sa pos)
+      else
+        defaultCase sa sb
+    _ -> defaultCase sa sb
+  where
+  defaultCase sa sb = do
+    markSimplification constToVal SimplifiableValue a sa
+    markSimplification constToVal SimplifiableValue b sb
+    pure (ValueSimp pos false t)
+
+lintValue env t@(Term (SubValue a b) pos) = do
+  sa <- lintValue env a
+  sb <- lintValue env b
+  case sa /\ sb of
+    (ConstantSimp _ _ v1 /\ ConstantSimp _ _ v2) -> pure (ConstantSimp pos true (v1 - v2))
+    (ConstantSimp _ _ v /\ _) ->
+      if (v == zero) then
+        pure (ValueSimp pos true (Term (NegValue b) pos))
+      else
+        defaultCase sa sb
+    (_ /\ ConstantSimp _ _ v) ->
+      if (v == zero) then
+        pure (simplifyTo sa pos)
+      else
+        defaultCase sa sb
+    _ -> defaultCase sa sb
+  where
+  defaultCase sa sb = do
+    markSimplification constToVal SimplifiableValue a sa
+    markSimplification constToVal SimplifiableValue b sb
+    pure (ValueSimp pos false t)
+
+lintValue env t@(Term (Scale (Term r@(Rational a b) pos2) c) pos) = do
+  sc <- lintValue env c
+  if (b == zero) then do
+    _ <- CMS.modify (over _warnings (Set.insert (DivisionByZero (termToRange r pos2))))
+    markSimplification constToVal SimplifiableValue c sc
+    pure (ValueSimp pos false t)
+  else
+    let
+      gcdv = gcd a b
 
-lintAction state (Term (Choice choiceId bounds) _) = over _holes (getHoles choiceId <> getHoles bounds) state
+      na = a `div` gcdv
 
-lintAction state (Term (Notify obs) _) = lintObservation state obs
+      nb = b `div` gcdv
 
-lintAction state hole@(Hole _ _ _) = over _holes (insertHole hole) state
+      isSimp = (abs gcdv) > one
+    in
+      case sc of
+        (ConstantSimp _ _ v) -> pure (ConstantSimp pos true (evalValue (makeEnvironment zero zero) (emptyState (Slot zero)) (S.Scale (S.Rational a b) (S.Constant v))))
+        (ValueSimp _ _ v) -> do
+          _ <-
+            if isSimp then
```

Comment:
I think you can use https://pursuit.purescript.org/packages/purescript-control/4.2.0/docs/Control.MonadZero#v:guard to make this a lot simpler although I'm not sure if it works with `State` or not

Reply:
hmm, I think `MonadZero` is missing from `Identity a`. This is a shame